File: DSA/Aarohi-Sharma/Day73_q1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int numberOfElement(TreeNode *root){
        if(root){
            return 1+ numberOfElement(root->left)+numberOfElement(root->right);
        }
        return 0;
    }
    bool btreeGameWinningMove(TreeNode* root, int n, int x) {
        if(root){
            if(root->val == x){
                int left = numberOfElement(root->left);
                int right = numberOfElement(root->right);
                int upper = n-left-right-1;
                if(upper>left+right+1||left>n/2 || right>n/2){
                    return true;
                }else{
                    return false;
                }
            }
            return btreeGameWinningMove(root->left,n,x)||btreeGameWinningMove(root->right,n,x);
        }
        return false;
    }
};

File: DSA/Ishita-Nigam/Day73/Day73_Q1.java
// Binary Tree Coloring Game
public class Day73_Q1 {

    static int xkaleft;
    static int xkaright;
    
    // TreeNode class definition (non-public)
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        // Constructor for TreeNode
        TreeNode(int val) {
            this.val = val;
            this.left = null;
            this.right = null;
        }
    }

    // Method to calculate the size of the tree and track the sizes of the left and right subtrees of node x
    public static int size(TreeNode node, int x) {
        if (node == null) {
            return 0;
        }

        int ls = size(node.left, x);
        int rs = size(node.right, x);

        int ts = ls + rs + 1;

        if (node.val == x) {
            xkaleft = ls;
            xkaright = rs;
        }
        return ts;
    }

    // Method to determine if the second player can win the game
    public boolean btreeGameWinningMove(TreeNode root, int n, int x) {
        size(root, x);

        int otherside = n - (xkaleft + xkaright + 1);
        int maxofthree = Math.max(otherside, Math.max(xkaleft, xkaright));

        int rest = n - maxofthree;

        return maxofthree > rest;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Example usage:
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        Day73_Q1 solution = new Day73_Q1();
        int n = 7; // Total number of nodes
        int x = 2; // Node selected by the first player

        boolean result = solution.btreeGameWinningMove(root, n, x);
        System.out.println("Can the second player win? " + result);
    }
}

